[Service Runtime] Add ability for service operations to be retried

ycombinator opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

This PR adds the ability for service operations to be retried.  Service components can optionally specify retries for their operations in their spec files like so:

```yml
version: 2
inputs:
  - name: endpoint
    service:
      operations:
        install:
          retry:
            init_interval: 60s
```

With the above spec definition, the `install` operation of the `endpoint` service will be retried indefinitely, with the duration between retries will start at 60 seconds and grow exponentially (with some random jitter). 

If `retry_sleep_init_duration` is omitted, it will default to **20 seconds**.

The duration between retries will grow exponentially (with some random jitter) but will be capped at approximately **15 minutes**.

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

## Why is it important?

In cases where a service operation may be flaky due to transient issues, it makes sense to retry the operation  indefinitely, giving it an opportunity to succeed.

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test


## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-2889/detail/PR-2889/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-2889/detail/PR-2889/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-2889/detail/PR-2889/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-2889/detail/PR-2889/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-agent_2889.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-06-16T04:35:38.191Z&rangeTo=2023-06-16T04:55:38.191Z&transactionName=BUILD+elastic-agent%2Felastic-agent-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=8535abfef49113806874f4796351e410&transactionId=f0c8752af6929a43)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-06-16T04:45:38.191+0000


* Duration: 23 min 1 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 6007  |
| Skipped      | 19 |
| Total        | 6026   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages.


- `run integration tests` : Run the Elastic Agent Integration tests.


- `run end-to-end tests` : Generate the packages and run the E2E Tests.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 98.667% (`74/75`) | :+1: 
Files | 68.605% (`177/258`) | :+1: 
Classes | 67.423% (`327/485`) | :+1: 
Methods | 53.846% (`1008/1872`) | :+1: 0.107
Lines | 39.669% (`11523/29048`) | :+1: 0.11
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

cmacknz commented on pkg/component/spec.go:
```diff
@@ -151,7 +151,9 @@ type ServiceOperationsSpec struct {
+	Args                   []string         `config:"args,omitempty" yaml:"args,omitempty"`
+	Env                    []CommandEnvSpec `config:"env,omitempty" yaml:"env,omitempty"`
+	Timeout                time.Duration    `config:"timeout,omitempty" yaml:"timeout,omitempty"`
+	RetryMaxCount          uint64           `config:"retry_max_count,omitempty" yaml:"retry_max_count,omitempty"`
```
In the case of endpoint security, wouldn't it make more sense to keep retrying indefinitely? We should always be trying to converge the state of running components to what it should be.

ycombinator replied:
++ to retrying indefinitely for service commands.  We could leave the `RetryMaxCount` setting here and service components could simply not specify it in their spec file to achieve infinite retries OR we could not even provide this setting.  Which one do you think is better?  I lean towards the latter as it reduces any opportunities for misconfiguration and also reduces the API surface; we can always add this setting back later if it becomes useful.

---

With retrying indefinitely, I think we need to do one of two other things:
* Either cap the interval between retries to X duration, as this interval will grow indefinitely (at an exponential rate), or
* Reset the interval between retries to the original one every X retries

Do you have a preference?  

---

Finally, I'm starting to wonder if this retry functionality for service commands is strictly or largely a property of the Agent itself (as opposed to the service component).  That is, does it even make sense to allow service components to define retry-related settings in their spec files?  

The only retry-related setting  that I could see service components wanting to control is perhaps `RetrySleepInitDuration` as that might depend on how long a service command might take to process under normal circumstances.

cmacknz replied:
> Which one do you think is better? I lean towards the latter as it reduces any opportunities for misconfiguration and also reduces the API surface;

Agreed, removing it is simpler with less chance of mistakes.

> Either cap the interval between retries to X duration, as this interval will grow indefinitely (at an exponential rate), or
> Reset the interval between retries to the original one every X retries

I would just cap the interval between retries. I think https://pkg.go.dev/github.com/cenkalti/backoff/v4#ExponentialBackOff configures the default `MaxInterval` to 15m which is probably reasonable if we can confirm that.

> The only retry-related setting that I could see service components wanting to control is perhaps RetrySleepInitDuration as that might depend on how long a service command might take to process under normal circumstances

Yeah we will want to allow configuring the minimum interval between retries, and also set it higher than the default. The main reason is to protect the host system from being overloaded. Lets image a worst case scenario where there is a bug in Elastic Defend, and it uses 90% of the host CPU and then fails installing after the 10 minute timeout elapses. If we retry too quickly we'll just be pointlessly wasting resources.

ycombinator replied:
> I would just cap the interval between retries. I think https://pkg.go.dev/github.com/cenkalti/backoff/v4#ExponentialBackOff configures the default `MaxInterval` to 15m which is probably reasonable if we can confirm that.
> 

The default `MaxInterval` used by https://pkg.go.dev/github.com/cenkalti/backoff/v4#ExponentialBackOff is `60s`.  

https://github.com/cenkalti/backoff/blob/a04a6fe64ffb0e3fd0816460529d300be5f252df/exponential.go#L89
https://github.com/cenkalti/backoff/blob/a04a6fe64ffb0e3fd0816460529d300be5f252df/exponential.go#L79

I'll up it to `15m` in our code.

ycombinator replied:
Still need to add a test case for execution succeeding after retrying once, but I've implemented the suggestions from this thread.

ycombinator commented on pkg/component/runtime/service_command.go:
```diff
@@ -55,24 +57,22 @@ func executeCommand(ctx context.Context, log *logger.Logger, binaryPath string,
+	go func() {
+		var errText string
+		scanner := bufio.NewScanner(ctxStderr)
+		for scanner.Scan() {
+			line := scanner.Text()
+			if len(line) > 0 {
+				txt := strings.TrimSpace(line)
+				if len(txt) > 0 {
+					errText = txt
+					// Log error output line
+					log.Error(errText)
 				}
 			}
-			errch <- errText
-		}()
-	}
+		}
+		errch <- errText
+	}()
```
This change is unrelated to the main changes in this PR.  It removes the `if ctxStderr != nil` conditional because the linter suggested that the condition would always evaluate to true.

ycombinator commented:
Looking into CI test failures on Windows...

ycombinator commented on pkg/component/runtime/service_command_test.go:
```diff
@@ -181,3 +195,162 @@ func TestExecuteCommand(t *testing.T) {
+		require.NoError(t, err)
+
+		// Give the command time to succeed.
+		time.Sleep(succeedCmdAfter)
```
Not a fan of this `time.Sleep` here but I haven't come up with a better solution without adding more complexity to the code path being tested.  Open to suggestions!

cmacknz replied:
Can it be replaced with a call to https://pkg.go.dev/github.com/stretchr/testify/require#Eventually?

blakerouse replied:
Yeah I agree an eventually check might be safer.

ycombinator replied:
I did try to use `require.Eventually` but couldn't come up with a good boolean condition for the check.  That's why I ended up resorting to a `time.Sleep` instead 😞.  Any ideas?

cmacknz replied:
You could write a variant of `checkRetryLogs` that polls for a log message indicating the service command executed successfully.

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

cmacknz commented on pkg/component/runtime/service_command.go:
```diff
@@ -93,11 +122,69 @@ func executeCommand(ctx context.Context, log *logger.Logger, binaryPath string,
 func executeServiceCommand(ctx context.Context, log *logger.Logger, binaryPath string, spec *component.ServiceOperationsCommandSpec) error {
+	// If there is a previous incarnation of the command executing (due to infinite
+	// retries), make sure to cancel it first before starting a new one here.
+	retryCancelFns.Cancel(binaryPath)
```
The command actually needs to be stopped before we can retry it. Is there a way we can confirm that?

https://github.com/elastic/elastic-agent/pull/2226 solved a previous version of this problem, where endpoint would be started before it was stopped and starting the second instance would always fail. There can only be a single instance of endpoint-security running on a machine at one time.

cmacknz replied:
You might need to synchronize on the output of the `Wait` call instead, all of the endpoint commands should end eventually there's just no real guarantee about when

https://github.com/elastic/elastic-agent/blob/7772cf2a42bd674a3347e04f9639d1b63ee9aabf/pkg/component/runtime/service_command.go#L77

The start process call is context aware, there's just no guarantee that the process is stopped immediately when the `cancel` function is executed. 

We end up calling `exec.CommandContext` which will call `os.Process.Kill`, but there's a chance it could happen after a small delay.

https://github.com/elastic/elastic-agent/blob/7772cf2a42bd674a3347e04f9639d1b63ee9aabf/pkg/core/process/cmd_darwin.go#L24

blakerouse commented on pkg/component/spec.go:
```diff
@@ -151,7 +151,8 @@ type ServiceOperationsSpec struct {
+	Args                   []string         `config:"args,omitempty" yaml:"args,omitempty"`
+	Env                    []CommandEnvSpec `config:"env,omitempty" yaml:"env,omitempty"`
+	Timeout                time.Duration    `config:"timeout,omitempty" yaml:"timeout,omitempty"`
+	RetrySleepInitDuration time.Duration    `config:"retry_sleep_init_duration,omitempty" yaml:"retry_sleep_init_duration,omitempty"`
```
Do you think this needs the `_init` in it? Could it just be called `retry_sleep_duration`?

Another idea is maybe to place the retry into its own structure, do we feel in the future we might add more retry options?

ycombinator replied:
> Do you think this needs the `_init` in it? Could it just be called `retry_sleep_duration`?

It's what the initial duration is.  After that, for subsequent retries, the duration will be increased each time.  Also, I was keeping the setting name consistent with another setting of the same name elsewhere: https://github.com/elastic/elastic-agent/blob/729636ac6a4d83286f4bfca3e8e0d2a2b1a46a6f/internal/pkg/agent/application/upgrade/artifact/config.go#L59

> Another idea is maybe to place the retry into its own structure, do we feel in the future we might add more retry options?

We could do this and/or change the name of the setting to remove `_init`, but I think we'll want to do the same thing in both places (for service command retries and artifact download retries) so there's consistency.  We released artifact download retries in 8.8.1 so it would be a breaking change to rename the setting there now?

faec replied:
Non-blocking alternate suggestion: what about the pattern used in many beats configs, `backoff { init, max }`? It's short and already understood by many users.

ycombinator replied:
> Non-blocking alternate suggestion: what about the pattern used in many beats configs, `backoff { init, max }`? It's short and already understood by many users.

Yeah, I think that's similar to what @blakerouse was suggesting with:

> Another idea is maybe to place the retry into its own structure, do we feel in the future we might add more retry options?

I'm open to doing this differently but the main "roadblock" I see is that we already have a setting elsewhere in Agent (for artifact download retries) that has the same meaning as this one here (for service command retries).

https://github.com/elastic/elastic-agent/blob/729636ac6a4d83286f4bfca3e8e0d2a2b1a46a6f/internal/pkg/agent/application/upgrade/artifact/config.go#L59

From a user's perspective, I think consistency is more important for settings in different places that have the same meaning? But maybe we could revisit both these settings (and any other similar places) in 9.0, when we could make breaking changes, and fix them all?  

On the other hand, the setting for the artifact download retries was introduced very recently - 8.8.1.  So maybe it's not too late to change it along with the setting here?

faec replied:
Hmm, possibly out of scope but I would be in favor of changing the artifact field names if it's still an option 😅

For the code under review specifically we do have a bit of extra leeway, since the spec behavior is only used internally by a known set of callers, so we can work around "breaking" changes if we really need to. Though on that note, please update the spec documentation in `docs/component-specs.md` with whatever field name/behavior you settle on.

ycombinator replied:
Yeah, good point that these settings are developer-facing vs. the artifact download ones, which are user-facing.  Given that, I'm +1 to changing the settings here to be something like: `retry { init }`, which should give us the ability to extend the retry settings in the future if needed.  

And yes, ++ to updating the spec documentation as well. 

I'll make both those changes in this PR.  Thanks!

blakerouse replied:
Just note we should push all the specification files to be owned by each components repository (instead of being inside of the Elastic Agent repository). As we make that change (not scheduled, but I think it something we should do, its not Elastic Agent job to know how to run the components), changes to the specification will be harder. So I am always in favor of ensure that the naming is easy to understand, clear, and can handle future changes where related settings might be added.

ycombinator replied:
I ended up going with `retry.init_interval` (as opposed to just `retry.init`) so we could, in the future, add settings like `retry.max_interval` or `retry.max_attempts`.

pierrehilbert commented:
Any objection to "backport' it to 8.9?

ycombinator commented:
@cmacknz @blakerouse I believe I've addressed all your feedback.  Please re-review when you get a chance.  Thanks!

blakerouse reviewed (approved):
Looks good, thanks for the fixes!

ycombinator commented:
Made a couple more changes to this PR:
- Simplified the logic with the `cmdDone` channel a bit; we are no longer passing it into `executeCommand`.
- Added a test to ensure that two successive calls to `executeServiceCommandWithRetries` will cancel the previous retries and command execution.

michalpristas commented on pkg/component/runtime/service_command.go:
```diff
@@ -93,11 +98,195 @@ func executeCommand(ctx context.Context, log *logger.Logger, binaryPath string,
+		binaryPath, spec.Args, envSpecToEnv(spec.Env), spec.Timeout,
+		retryCtx, retrySleepInitDuration, retrySleepMaxDuration,
+	)
+	return nil
```
no need to have error return if we're not returning error in any case

ycombinator replied:
Removed.

michalpristas commented on pkg/component/runtime/service_command.go:
```diff
@@ -93,11 +98,195 @@ func executeCommand(ctx context.Context, log *logger.Logger, binaryPath string,
+}
+
+func (cr *cmdRetrier) Stop(cmdKey uint64, log *logger.Logger) {
+	cr.mu.RLock()
```
you're modifying cr.cmds here, should be a proper lock. 
Stop is not used elsewhere can be unexported and merged with untrack as it seems its idempotent

ycombinator replied:
Yes, good catch, thanks.  A previous version of this method was only reading from `cr.cmds`; I forgot to update the lock after that changed.

ycombinator replied:
Fixed.

ycombinator commented on pkg/component/runtime/service_command.go:
```diff
@@ -93,11 +98,195 @@ func executeCommand(ctx context.Context, log *logger.Logger, binaryPath string,
+	digest := fnv.New64()
+	digest.Write([]byte(sb.String()))
+
+	return digest.Sum64()
```
> looking at the fnv1 and fnv1a there's a high probability of collisions, how do we want to handle that?

Where are you seeing this? 
 I looked at https://github.com/Cyan4973/xxHash/wiki/Collision-ratio-comparison and fnv looked decent to me.  I've used `xxHash` in the past but I didn't want to pull in another dependency just for the usage here.  Do you have any recommendation for an alternate hashing algorithm that has a lower probability of collisions than fnv?

michalpristas replied:
https://theses.liacs.nl/pdf/2014-2015NickvandenBosch.pdf

michalpristas replied:
i'm just thinknig how many concurrent intalls/uninstalls can one have. under any circumestantces there should not be different versions of service being installed so spec args are the same for service type always. maybe we dont even need this type of hashing. if new install is starting it will be install of the exactly same service as the one already running.

ycombinator replied:
The issue is that at this level of the code path, we don't know which service operation (`install`, `uninstall`) is being run.  So it's possible that the command for the `install` operation is already running when the command for `uninstall` is called.  In this scenario, I think we'd want the running `install` operation command to finish, that is, we would not want to stop it.  So I think, to be safe, it's better to have the hashing or some other way to know that the command being run is exactly the same as a command already running or not.

Thanks for sharing the paper.  Looking at the table in there, I think SHA256 with 64-bit output should suffice.

michalpristas replied:
i mean yes that's true, but it costs us a few cycles. but this not being a hot path i think we should be fine

ycombinator commented on pkg/component/runtime/service_command.go:
```diff
@@ -93,11 +98,195 @@ func executeCommand(ctx context.Context, log *logger.Logger, binaryPath string,
 }
 
 func executeServiceCommand(ctx context.Context, log *logger.Logger, binaryPath string, spec *component.ServiceOperationsCommandSpec) error {
+	return executeServiceCommandWithRetries(
```
Fixed.

ycombinator commented:
/test

ycombinator commented:
/test

ycombinator commented:
/test

ycombinator commented:
/test

ycombinator commented:
That's three successful builds (with no changes) in a row.  I'm going to assume the `TestExecuteServiceCommand/succeed_after_retry` test is no longer flaky and will merge this PR now.
